test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/TransportLayerTests/RequestMessageArgsTests.cs

```diff
@@ -0,0 +1,121 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="RequestMessageArgsTests.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved.
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using Microsoft.AspNetCore.OData;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Client.E2E.Tests.TransportLayerTests.Server;
+using Microsoft.OData.E2E.TestCommon;
+using Microsoft.OData.E2E.TestCommon.Common.Client.EndToEnd.Default;
+using Microsoft.OData.E2E.TestCommon.Common.Server.EndToEnd;
+using Xunit;
+using ClientEndToEndModel = Microsoft.OData.E2E.TestCommon.Common.Client.EndToEnd;
+
+namespace Microsoft.OData.Client.E2E.Tests.TransportLayerTests
+{
+    public class RequestMessageArgsTests : EndToEndTestBase<RequestMessageArgsTests.TestsStartup>
+    {
+        private readonly Uri _baseUri;
+        private readonly Container _context;
+        private IODataRequestMessage lastRequestMessage = null;
+
+        public class TestsStartup : TestStartupBase
+        {
+            public override void ConfigureServices(IServiceCollection services)
+            {
+                services.ConfigureControllers(typeof(RequestMessageArgsTestsController), typeof(MetadataController));
+
+                services.AddControllers().AddOData(opt => opt.Count().Filter().Expand().Select().OrderBy().SetMaxTop(null)
+                    .AddRouteComponents("odata", CommonEndToEndEdmModel.GetEdmModel()));
+            }
+        }
+
+        public RequestMessageArgsTests(TestWebApplicationFactory<TestsStartup> fixture)
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+
+            _context = new Container(_baseUri)
+            {
+                HttpClientFactory = HttpClientFactory
+            };
+
+            ResetDataSource();
+        }
+
+        [Fact]
+        public async Task AddingCustomHeaderToRequest_ShouldSucceed()
+        {
+            string headerName = "MyNewHeader";
+            string headerValue = "MyNewHeaderValue";
+
+            Func<DataServiceClientRequestMessageArgs, DataServiceClientRequestMessage> configureRequest =
+                args =>
+                {
+                    args.Headers.Add(headerName, headerValue);
+                    return new HttpClientRequestMessage(args);
+                };
+
+            this.UpdateContext(configureRequest);

```

Comment:
Since the core of what we're testing as that custom configuration of the request message is honored by the `DataServiceContext`, I think it's important that we use show the same methods in the test that the user would use in real life to achieve that, instead of relying on this `UpdateContext` method that obscures what the test is doing. It also makes the test harder to understand and review.